### ISSUE TITLE
There are some transparent points in lines drawed by quadraticCurveTo

### ISSUE THREAD

**Issue by daigocy** [pixijs/graphics-smooth#13]
code is 
```javascript
        ins.clear();
        for (let i = 0; i < path.length; i += 4) {
            ins.lineStyle({
                width,
                color,
                cap: PIXI.LINE_CAP.ROUND,
                join: PIXI.LINE_JOIN.ROUND,
            });
            ins.moveTo(path[i - 2] || 0, path[i - 1] || 0);
            ins.quadraticCurveTo(path[i], path[i + 1], path[i + 2], path[i + 3]);
        }
```
There are some transparent points in lines drawed by quadraticCurveTo.

**Comment by dev7355608** [comment]
Fixed by dba5d581e5b54a9d114d1194af108b45d7f87382 (v0.0.21).

@ivanpopelyshev I assume the `abs(D) < 0.01` branch was added for performance reasons. The `0.01` threshold was too high; in consequence line segments didn't properly connect if the angle is tiny. I simply changed it to `D == 0.0`, instead of lowering the threshold, which would need to depend on `shift + dy`. So now those artifacts should be impossible under any circumstances. Because `D` is very rarely 0, we could remove this branch entirely, or, if it's important for performance, try to find a better threshold.

Example image that shows the transparent pixels:
![image](https://user-images.githubusercontent.com/31905376/155023266-0e0ed34b-8e54-435b-84c2-52bb3573ca0b.png)

**Comment by ivanpopelyshev** [comment]
Yep, something like that. need better eps maybe

**Comment by daigocy** [comment]
Problem become more serious
![企业微信截图_16455817318659](https://user-images.githubusercontent.com/29224385/155250330-5ee221fb-c3c8-4ffb-bb6b-d4b544e12fc6.png)

**Comment by daigocy** [comment]
> Fixed by [dba5d58](https://github.com/pixijs/graphics-smooth/commit/dba5d581e5b54a9d114d1194af108b45d7f87382) (v0.0.21).
> 
> @ivanpopelyshev I assume the `abs(D) < 0.01` branch was added for performance reasons. The `0.01` threshold was too high; in consequence line segments didn't properly connect if the angle is tiny. I simply changed it to `D == 0.0`, instead of lowering the threshold, which would need to depend on `shift + dy`. So now those artifacts should be impossible under any circumstances. Because `D` is very rarely 0, we could remove this branch entirely, or, if it's important for performance, try to find a better threshold.
> 
> Example image that shows the transparent pixels: ![image](https://user-images.githubusercontent.com/31905376/155023266-0e0ed34b-8e54-435b-84c2-52bb3573ca0b.png)

**Comment by dev7355608** [comment]
@daigocy Could you share the code that produces these artifacts? I haven't see those in my tests.

**Comment by daigocy** [comment]
> @daigocy Could you share the code that produces these artifacts? I haven't see those in my tests.

I get points from mousemove event and calculate path from points.
```typescript
_update_points(points: number[]) {
        const path: number[] = [];
        const lastPoint = {
            x: 0,
            y: 0,
        };
        for (let i = 0; i < points.length; i += 2) {
            path.push(lastPoint.x, lastPoint.y, (lastPoint.x + points[i]) / 2, (lastPoint.y + points[i + 1]) / 2);
            lastPoint.x = points[i];
            lastPoint.y = points[i + 1];
        }
        this.points = points;
        this.path = path;
        this.updateBorderRect(points);
    }
```
then use path to draw lines drawed by quadraticCurveTo.
```typescript
function drawLine(ins: Graphics, props: LineProps) {
    const { lineWidth: width, color, path, notEnded } = props;
    ins.clear();
    if (path.length) {
        for (let i = 0; i < path.length; i += 4) {
            ins.lineStyle({
                width,
                color,
                cap: PIXI.LINE_CAP.ROUND,
                join: PIXI.LINE_JOIN.ROUND,
            });
            ins.moveTo(path[i - 2] || 0, path[i - 1] || 0);
            ins.quadraticCurveTo(path[i], path[i + 1], path[i + 2], path[i + 3]);
        }
    } else if (!notEnded) {
        ins.moveTo(0, 0);
        ins.beginFill(color);
        ins.drawCircle(0, 0, width / 2);
        ins.endFill();
    }
}
```

**Comment by daigocy** [comment]
> @daigocy Could you share the code that produces these artifacts? I haven't see those in my tests.

I see your code in [dba5d58](https://github.com/pixijs/graphics-smooth/commit/dba5d581e5b54a9d114d1194af108b45d7f87382) and  resume the changes in line 252 and line 309 then it works better. But I down known the reason.

**Comment by dev7355608** [comment]
@daigocy Could you provide any path array that has these artifacts as well please.

**Comment by daigocy** [comment]
> @daigocy Could you provide any path array that has these artifacts as well please.

```typescript
const lineWidth = 10;
const points = [8, 0, 28, 2, 63, 11, 94, 18, 126, 28, 146, 38, 161, 47, 170, 56, 174, 68, 174, 80, 174, 92, 170, 100, 152, 113, 134, 117, 105, 123, 77, 124, 56, 127, 36, 130, 24, 132, 19, 132, 17, 133, 16, 133, 16, 133]
const path = [0, 0, 4, 0, 8, 0, 18, 1, 28, 2, 45.5, 6.5, 63, 11, 78.5, 14.5, 94, 18, 110, 23, 126, 28, 136, 33, 146, 38, 153.5, 42.5, 161, 47, 165.5, 51.5, 170, 56, 172, 62, 174, 68, 174, 74, 174, 80, 174, 86, 174, 92, 172, 96, 170, 100, 161, 106.5, 152, 113, 143, 115, 134, 117, 119.5, 120, 105, 123, 91, 123.5, 77, 124, 66.5, 125.5, 56, 127, 46, 128.5, 36, 130, 30, 131, 24, 132, 21.5, 132, 19, 132, 18, 132.5, 17, 133, 16.5, 133, 16, 133, 16, 133]
```

**Comment by dev7355608** [comment]
Looks good to me. How does it look on your machine?
![image](https://user-images.githubusercontent.com/31905376/155258158-7678f0c0-2865-4977-bef9-84245e2c2d65.png)
```js
function drawLine(ins, props) {
    const { lineWidth: width, color, path, notEnded } = props;
    ins.clear();
    if (path.length) {
        for (let i = 0; i < path.length; i += 4) {
            ins.lineStyle({
                width,
                color,
                cap: PIXI.LINE_CAP.ROUND,
                join: PIXI.LINE_JOIN.ROUND,
            });
            ins.moveTo(path[i - 2] || 0, path[i - 1] || 0);
            ins.quadraticCurveTo(path[i], path[i + 1], path[i + 2], path[i + 3]);
        }
    } else if (!notEnded) {
        ins.moveTo(0, 0);
        ins.beginFill(color);
        ins.drawCircle(0, 0, width / 2);
        ins.endFill();
    }
}

const g = new PIXI.smooth.SmoothGraphics()
const lineWidth = 10;
const points = [8, 0, 28, 2, 63, 11, 94, 18, 126, 28, 146, 38, 161, 47, 170, 56, 174, 68, 174, 80, 174, 92, 170, 100, 152, 113, 134, 117, 105, 123, 77, 124, 56, 127, 36, 130, 24, 132, 19, 132, 17, 133, 16, 133, 16, 133]
const path = [0, 0, 4, 0, 8, 0, 18, 1, 28, 2, 45.5, 6.5, 63, 11, 78.5, 14.5, 94, 18, 110, 23, 126, 28, 136, 33, 146, 38, 153.5, 42.5, 161, 47, 165.5, 51.5, 170, 56, 172, 62, 174, 68, 174, 74, 174, 80, 174, 86, 174, 92, 172, 96, 170, 100, 161, 106.5, 152, 113, 143, 115, 134, 117, 119.5, 120, 105, 123, 91, 123.5, 77, 124, 66.5, 125.5, 56, 127, 46, 128.5, 36, 130, 30, 131, 24, 132, 21.5, 132, 19, 132, 18, 132.5, 17, 133, 16.5, 133, 16, 133, 16, 133]

drawLine(g, {lineWidth, color: 0xff0000, path, notEnded: true});
```

**Comment by daigocy** [comment]
> ```js
> , 
> ```

![image](https://user-images.githubusercontent.com/29224385/155259740-4e4ad631-db4c-441c-8a57-8c7c52b0d920.png)

**Comment by daigocy** [comment]
My web demo is https://wwwr.plaso.cn/static/sdk/websdkdemo/board.html?env=dev&loginName=editor_0&ownerLoginName=cy&recordId=x22&signature=7DEA2F034223006049DF120ED70DC5523F7551AA&userType=editor&validBegin=1645594255&validTime=10800

**Comment by mudcube** [comment]
@daigocy Demo looks lovely to me, no issues on my machine :)

**Comment by daigocy** [comment]
I write a simple demo  using pixi.js@^6.2.0. 
```javascript
import * as PIXI from "pixi.js";
import { SmoothGraphics } from '@pixi/graphics-smooth';
window.PIXI = PIXI;

const app = new PIXI.Application({antialias: true});

// The application will create a canvas element for you that you
// can then insert into the DOM
document.body.appendChild(app.view);

function drawLine(ins, props) {
    const { lineWidth: width, color, path, notEnded } = props;
    ins.clear();
    if (path.length) {
        for (let i = 0; i < path.length; i += 4) {
            ins.lineStyle({
                width,
                color,
                cap: PIXI.LINE_CAP.ROUND,
                join: PIXI.LINE_JOIN.ROUND,
            });
            ins.moveTo(path[i - 2] || 0, path[i - 1] || 0);
            ins.quadraticCurveTo(path[i], path[i + 1], path[i + 2], path[i + 3]);
        }
    } else if (!notEnded) {
        ins.moveTo(0, 0);
        ins.beginFill(color);
        ins.drawCircle(0, 0, width / 2);
        ins.endFill();
    }
}

const g = new SmoothGraphics()
const lineWidth = 10;
const points = [8, 0, 28, 2, 63, 11, 94, 18, 126, 28, 146, 38, 161, 47, 170, 56, 174, 68, 174, 80, 174, 92, 170, 100, 152, 113, 134, 117, 105, 123, 77, 124, 56, 127, 36, 130, 24, 132, 19, 132, 17, 133, 16, 133, 16, 133]
const path = [0, 0, 4, 0, 8, 0, 18, 1, 28, 2, 45.5, 6.5, 63, 11, 78.5, 14.5, 94, 18, 110, 23, 126, 28, 136, 33, 146, 38, 153.5, 42.5, 161, 47, 165.5, 51.5, 170, 56, 172, 62, 174, 68, 174, 74, 174, 80, 174, 86, 174, 92, 172, 96, 170, 100, 161, 106.5, 152, 113, 143, 115, 134, 117, 119.5, 120, 105, 123, 91, 123.5, 77, 124, 66.5, 125.5, 56, 127, 46, 128.5, 36, 130, 30, 131, 24, 132, 21.5, 132, 19, 132, 18, 132.5, 17, 133, 16.5, 133, 16, 133, 16, 133]

drawLine(g, {lineWidth, color: 0xff0000, path, notEnded: true});
g.x = 100;
g.y = 100;

app.stage.addChild(g);
```
This is result
![企业微信截图_16455966245509](https://user-images.githubusercontent.com/29224385/155269546-b5def4e1-1ff2-4986-bada-29b6f8ccdd4a.png)

**Comment by daigocy** [comment]
> Looks good to me. How does it look on your machine? ![image](https://user-images.githubusercontent.com/31905376/155258158-7678f0c0-2865-4977-bef9-84245e2c2d65.png)
> 
> ```js
> function drawLine(ins, props) {
>     const { lineWidth: width, color, path, notEnded } = props;
>     ins.clear();
>     if (path.length) {
>         for (let i = 0; i < path.length; i += 4) {
>             ins.lineStyle({
>                 width,
>                 color,
>                 cap: PIXI.LINE_CAP.ROUND,
>                 join: PIXI.LINE_JOIN.ROUND,
>             });
>             ins.moveTo(path[i - 2] || 0, path[i - 1] || 0);
>             ins.quadraticCurveTo(path[i], path[i + 1], path[i + 2], path[i + 3]);
>         }
>     } else if (!notEnded) {
>         ins.moveTo(0, 0);
>         ins.beginFill(color);
>         ins.drawCircle(0, 0, width / 2);
>         ins.endFill();
>     }
> }
> 
> const g = new PIXI.smooth.SmoothGraphics()
> const lineWidth = 10;
> const points = [8, 0, 28, 2, 63, 11, 94, 18, 126, 28, 146, 38, 161, 47, 170, 56, 174, 68, 174, 80, 174, 92, 170, 100, 152, 113, 134, 117, 105, 123, 77, 124, 56, 127, 36, 130, 24, 132, 19, 132, 17, 133, 16, 133, 16, 133]
> const path = [0, 0, 4, 0, 8, 0, 18, 1, 28, 2, 45.5, 6.5, 63, 11, 78.5, 14.5, 94, 18, 110, 23, 126, 28, 136, 33, 146, 38, 153.5, 42.5, 161, 47, 165.5, 51.5, 170, 56, 172, 62, 174, 68, 174, 74, 174, 80, 174, 86, 174, 92, 172, 96, 170, 100, 161, 106.5, 152, 113, 143, 115, 134, 117, 119.5, 120, 105, 123, 91, 123.5, 77, 124, 66.5, 125.5, 56, 127, 46, 128.5, 36, 130, 30, 131, 24, 132, 21.5, 132, 19, 132, 18, 132.5, 17, 133, 16.5, 133, 16, 133, 16, 133]
> 
> drawLine(g, {lineWidth, color: 0xff0000, path, notEnded: true});
> ```

My friend use safari and demo looks ok. I am using chrome.

**Comment by dev7355608** [comment]
@daigocy Could you check whether this is related to just POINT line joints/caps or others as well? Set the line alpha to 0.5: that might help to see what goes wrong.

**Comment by daigocy** [comment]
@dev7355608 Now I set the line alpha to 0.5 and remove  "ins.moveTo(path[i - 2] || 0, path[i - 1] || 0);" from my code. 
When I set lineCap to PIXI.LINE_CAP.BUTT or PIXI.LINE_CAP.SQUARE, no gap appears in the line.
But When I set lineCap to PIXI.LINE_CAP.ROUND, sometimes gap will appear at the start or the end of the line.
![企业微信截图_16456815258072](https://user-images.githubusercontent.com/29224385/155465477-5c248540-68b8-41af-a8a4-1810b0eaed3f.png)

**Comment by dev7355608** [comment]
The initial fix introduced potential precision/divide-by-zero problems with collinear edges. I addressed that in a1754e7cc7c00329634a3ec65a9ff4d3ffe653ac (v0.0.22). I'm not sure that's it, so let me know @daigocy.

**Comment by daigocy** [comment]
@dev7355608 It's perfect now. Thank you.
![4e3ac75e-162a-4d84-acc4-416543a04928](https://user-images.githubusercontent.com/29224385/155677464-0266346e-530b-4922-b096-b55e09975621.png)
